### ISSUE TITLE
autorun( );

### ISSUE THREAD

**Issue by E3t1jump** [Engineering-3/RCM-BFin#1]
We need to verify if we have the correct instructions for autorun( );

xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
First line of the program needs to be 

autorun(5);

Then upload your program to flash sector 4.    Next time you boot, it will
load and start the program.  Once it starts, you have 5 seconds to send an
ESC character to break execution and regain control from the console.

To clear Sector 4

From a console, 'zc' clears the flash buffer and 'zw' then writes the
empty flash buffer to sector 4, thereby clearing the sector.  Via the
admin page, you could just write a file with blanks to sector 4.

**Comment by EmbeddedMan** [comment]
I should be able to figure out what we've got working right now pretty easily. I'll report back when I have more details.

**Comment by E3t1jump** [comment]
Additional Info: As bits of my memory are coming back on this, Sector 4 is where autorun( ); programs need to start, but they can fill in multiple continuous sectors (i.e. sectors 4-10/other).

**Comment by EmbeddedMan** [comment]
OK, here's what I've learned from reading up in the code, docs, and playing around.

The basic idea of autorun is working. If you load a PicoC program with autorun(x); as the first line (where x is the number of seconds you want the BFin to wait on boot before running the next lines of the program), and then send a zW04 command to write what's in the RAM buffer into sector 4, then your program will run every time the BFin is booted.

Yea. 

The problem, however, is that MC has no idea that the BFin is "in PicoC mode" at boot. So when you reboot the BFin, the PicoC program does run, but you don't see anything in the output window in MC, since MC doesn't know to be asking the BFin for PicoC output packets.

You can prove this to yourself by doing the following:
1) As per above, copy a simple PicoC program into sector 04 - say one that prints out a bunch of stuff. Make sure the program takes at least several seconds to run.
2) Now click Run Code in MC. But before the program has a chance to finish, reset the BFin with the shorting jumper reset. This will keep MC thinking that the BFin is in PicoC mode, and the Bfin will run the autostart PicoC program.
3) You'll see some 'raw' packet data after a few seconds, which is the output of the autorun(x); command being interpreted as PicoC output data packets and displayed in the output window.
4) Then you'll see the output of your PicoC program appear,.
5) The problem now is that because we didn't run PicoC 'normally' (i.e. it was autorun at boot) we didn't print out "Leaving PicoC" and thus MC doesn't know to leave PicoC mode automatically. A click of the ESC button fixes that. 

So, how best to fix the situation? Not sure yet. I'm thinking about it. Should we have MC continually polling for PicoC reply packets, and just get rid of the whole idea of "PicoC mode"?

Or should we maybe have the BFin send a new kind of 'reply' packet when it boots up and finds a PicoC program to autorun? This new packet type would get sent to MC, and tell it "hey, I just booted, and I'm running a PicoC program", which would then put MC into PicoC mode so it would get the PicoC output?

Or something else?

**Comment by E3t1jump** [comment]
A couple of thoughts and questions:

Where we used autorun( ); in the past was when our robot had to boot into the operations code (no communication with the robot allowed) in RoboCup Jr. We used this the way code is stored/run in the original LEGO RCX (only we just had the one code sector while the RCX had five).

My current thinking is, if we need to boot in a Ready to Go format, something like the code to activate the IMU may work well as an autorun( );. This looks like it could be a nice Start feature, but we need some way to utilize it through MainController if possible.

A) Can we direct code to load starting in sector 4 of the BFin RAM buffer through MC (I only remember doing this through a terminal or JTAG)? I'm not sure how you are sending the zW04 command through MC and PicoC Load Code.

B) Should we add a special Load Code option in MC that is specific for autorun( );?

C) If we have a special designation for loading code into autorun( ); would adding an identifier here that we have an autorun( ); code be a way to notify MC to look for this somehow?

D) How do we access the ESC feature of autorun ( ): to stop the code from running within the delay period set in the autorun( ); argument? (Or do we??? Maybe all autorun( ); code starts with a wait loop such that a button needs to be triggered or the code does not proceed. ESC in MC can then be used to get out of the code?)

E) Do we need autorun( ); if we can use PicoC shared variables and interactive mode to run pre-set codes by the push of a button?

**Comment by EmbeddedMan** [comment]
Autorun is completely available, today, from within MC. You just have to enter the commands as I specified above in the "DIrect command to send" field.

Now, it wouldn't hurt to have a new button labeled "copy PicoC to Autorun sector", and another labeled "clear Autorun sector" or something like that. Instead of having to enter the commands by hand, these buttons would do it for you.

The 'ESC' functionality is not working in a way that's useful to users at the moment during the autorun(x); command. This would have to be something we'd fix. I'm not sure what the best way is to do that yet. 

Right now MC doesn't have a way of detecting that the BFin's been rebooted. When the BFin is rebooted, it send a Version string to MC, but not in packet format (so that we remain compatible with older clients who don't use packet format), and we also send "Checking for autorun()" but again not in packet format. Maybe that should change. Maybe we should start out, from the very beginning in packet mode, and then the 'esc' during autorun() would work. I'll need to try that. 

As for E) above, I think they serve two very different purposes. You'll need to decide if the PicoC shared variable function gives you everything you need (from autorun) or if you want them both to survive.

**Comment by E3t1jump** [comment]
To clarify the process for loading autorun( ); code:

A) Write a code as normal, but have the first line of the code be the autorun(x): with the value of x as the delay.

B) Use the Load Code button to load the code into RAM as normal.

C) Now, send the zW04 command through the Direct Command to Send window to write the code from RAM into the non-volatile memory starting at sector 04.

**Comment by EmbeddedMan** [comment]
Exactly correct.

**Comment by E3t1jump** [comment]
Before we launch into greater detail in autorun( ); let's clarify how to use the PicoC shared variable and interactive mode so we (I) have a fuller picture of what to envision for using each option.

**Comment by E3t1jump** [comment]
This is where we stand on autorun( );

A) We can load autorun( ); programs through MC:
          1) Write a code as normal, but have the first line of the code be the autorun(x): with the value of 
               x as the delay.
          2) Use the Load Code button to load the code into RAM as normal.
          3) Now, send the zW04 command through the Direct Command to Send window to write the code 
              from RAM into the non-volatile memory starting at sector 04.

B) We cannot activate Escape through MC for an autorun( ): program (without a significant 'x' value/delay... this kind of defeats the purpose of autorun( );) as it takes time to connect to the RCM with MC after the robot boots. Considerations:
          1) We can add a touch sensor/other to the RCMx1 to act as the Escape/Exit trigger. 
          2) Code added to the autorun( ); program can be used as an Exit.
          3) Consider a hard code in BFin firmware to accommodate the Escape prior to autorun( ); launch.

C) Per the discussion in an earlier comment (re: MC does not detect reboot as the reboot elements are not in packet mode AND should we start out from the very beginning in packet mode): Because of the delay between RCM boot and connecting with MC, is there any advantage to changing to always being in packet mode (especially related to the autorun( ); feature)?

**Comment by E3t1jump** [comment]
Note: If the BFin is rebooted by shorting the boot pins, this breaks the streaming function from sensors (sensors set to streaming stop updating) and requires MC be Stopped/reRun to regain data streaming. The sensors are still active, its just their data stops updating in MC SensorView.

**Comment by EmbeddedMan** [comment]
About C) above : I believe we already are in packet mode right after reboot. I might be wrong, but I thought that's  how it's set up right now. I don't see why MC couldn't detect a reboot of the BFin - we just have to make sure the BFin sends out it's version string every time it boots (I think it's already doing this) and make sure MC is ready to receive it at all times (which it may already be set up for). 

As a separate issue, we could have MC add the ability to detect a BFin reboot and re-send all sensor configuration at that time. Might be nice.

Even with MC detecting reboots, hitting the ESC button quickly enough might be a challenge. While it would be great for MC to have the ability to stop an autorun() script, I don't think you want to count on it as your only way of breaking out of the autorun.

**Comment by E3t1jump** [comment]
After reboot, motor controls and video, loading and running programs all work as normal. It is just the sensors set for streaming mode stop updating. I am not seeing a version return after reboot.

I'll make a new issue for this as it would be good for MC to fully kick it after a reboot, just let me know if it is a MC issue only, or if it needs threads in both MC and BFin.

As for the ESC for autorun( ); I see this as outside of the MC functions as the BFin goes into a non-boot state during the autorun( ); delay and there is no connection with MC during this delay period.

**Comment by EmbeddedMan** [comment]
After checking on my BFin, I see that I was slightly wrong.

The BFin does in fact send out it's version string every boot, without prompting. However, it does so without having the version string inside a packet. So you were right - the BFin starts out NOT in packet mode at boot. 

I'm going to try changing that, just to see if we can get MC to recognize the version string being sent out on reboot.

**Comment by EmbeddedMan** [comment]
Yup, that works just great. So if we start off in packet mode from boot, MC will see our reboot message. We can then have MC trigger off that to 'detect a reboot' and thus re-send the sensor configuration down.

In addition, MC will need to be changed so that after it uploads new firmware, it does not expect the BFin to leave packet mode (Like it does now), but rather stay in packet mode the whole time.

**Comment by EmbeddedMan** [comment]
In addition, we want to be able to process normal commands while the autorun() delay is happening. I will make that change to PicoC as well.
